src/serde/de.rs

```diff
@@ -526,25 +521,47 @@ impl<'de> de::MapAccess<'de> for MapDeserializer<'_> {
                     self.value = Some(value);
                     let visited = Rc::clone(&self.visited);
                     let key_de = Deserializer::from_parts(key, self.options, visited);
-                    return seed.deserialize(key_de).map(Some);
+                    return Ok(Some(key_de));
                 }
                 None => return Ok(None),
             }
         }
     }
 
-    fn next_value_seed<T>(&mut self, seed: T) -> Result<T::Value>
-    where
-        T: de::DeserializeSeed<'de>,
-    {
+    fn next_value_deserializer(&mut self) -> Result<Deserializer> {
         match self.value.take() {
             Some(value) => {
                 let visited = Rc::clone(&self.visited);
-                seed.deserialize(Deserializer::from_parts(value, self.options, visited))
+                Ok(Deserializer::from_parts(value, self.options, visited))
             }
             None => Err(de::Error::custom("value is missing")),
         }
     }
+}
+
+impl<'de> de::MapAccess<'de> for MapDeserializer<'_> {
+    type Error = Error;
+
+    fn next_key_seed<T>(&mut self, seed: T) -> Result<Option<T::Value>>
+    where
+        T: de::DeserializeSeed<'de>,
+    {
+        match self.next_key_deserializer() {
+            Ok(Some(key_de)) => seed.deserialize(key_de).map(Some),
+            Ok(None) => Ok(None),
+            Err(error) => Err(error),
+        }
```

Comment:
If you prefer I can make this more idiomatic, but in functions that get instantiated a lot, I prefer to keep it simple to keep the generated lines minimal.